### DISCUSSION TITLE
Modbus RTU : Confusing and help on wich library to use

### DISCUSSION THREAD

**Discussion by nicolouts** [eModbus/eModbus#376] [eModbus applications]
Hello all,

I want to develop an application with an ESP32 which cyclically interrogates several devices with the RTU protocol.
But I'm a little confused about what terms to use, Client-Server-Master-Slave.

My goal is to use an ESP32 to request data for example to a device with ID 1, after data to device with ID 2, after data to device with ID 3, etc and then restart to asking data to device ID 1.

Can you explain me wich library I need to use? Because I'm confuse.
I already have use ModbusClientRTU.h/ModbusMessage.h, I can read data from device ID 1 but after I receive a timeout when asking data to device ID 2.
If I take each device separately in a program, I can communicate with it. But if I want to put them both in the same program, I have a timeout on device id 2.

What is the best way to ask data to different device, put request to the queue and wait for feedback or ask data to device id 1 then wait for answer device ID 1 and then device ID 2.

I have set the device baud rate to 115200, but it look like between request and answer, I have around 9ms.
Is it pssible to reduce this time?

Thank you in advance

Nicolouts

**Comment by Miq1** [comment]
Welcome in the beautiful world of Modbus RTU 😆 

What you want to achieve is a very common application and should be easy to implement. 

To first sort out the names of the Modbus devices: a "client" was formerly called master - that has been changed in the Modbus standard - and is the device sending  requests on the bus. A "server" formerly was a slave; these devices will answer to requests of the client with responses.

In short you will have to
- ``#include <ModbusClientRTU.h>`` as you already did. No other includes are necessary for an eModbus RTU client.
- be sure the servers are configured to have different server IDs. This normally is done in a menu of the server, by DIP switsches on the board or anything other. These server IDs are needed to address that very server.
- in your program implement both a data handler and an error handler, as can be seen in the examples. The former is to receive and process regular responses, the latter to catch and handle error messages the servers may have sent.
- In your main loop you can determine the interval to send requests. Requests are always issued by the ``addRequest()`` method of the client object. They will have the server ID of the respective server the request is to be addressed to.
- In your handlers you can check the server ID of the response or error message to see which server was the sender.

The requests will be sent in strictly that order in that you added them. You can add requests as quickly as you like, but they will be serialized anyway. The handler functions you provided will be called asynchronously as soon as a response has been received. The common procedure is to store away the data read from the response and set a flag. This flag will be checked in the main loop and data processing will take place there. This is done to not block the eModbus task (running on another core) with too much processing in the handler functions (that are run by the eModbus core!).

On Modbus RTU there is a mandatory "quiet time" between messages, based on the transmission speed, but no less than 1.7ms. For a request/response pair you will have that interval twice, so whatever you do you will never get below 3.5ms of turnaround time, even with no time for processing request and response.

If your devices will support that, the ESP32 with eModbus is able to go up to 5 Mbaud by the way.

**Reply by nicolouts** [reply]
Hello Miq1,

Thank you for your fast feedback.
Nox, its a little more clear 😆

I have done a test code with your recommendation, but I can't read data from device ID2. I always receive a timeout answer.
Could you please help me?

Thank you

Nicolouts

Here is my code

```
#include <Arduino.h>

//#include <ESP_Panel_Library.h>
//#include <ESP_IOExpander_Library.h>

//RTC
#include "PCF85063A.h"

#include "ModbusClientRTU.h"
#include "Logging.h"

#define MODBUS_RX       43
#define MODBUS_TX       44
#define MODBUS_BAUD     115200UL
#define MODBUS_CONFIG   SERIAL_8N1
#define MODBUS_SERIAL   Serial2

//ModbusMaster object
// Create a ModbusRTU server
ModbusClientRTU MBClient;

Error MBError;
uint32_t MBToken;
bool data_request;
bool data_ready;
uint8_t request;
Error err;
uint16_t values[8];

static uint32_t next_request;


static datetime_t Set_Time = {
    .year = 2024,
    .month = 02,
    .day = 02,
    .dotw = 5,
    .hour = 9,
    .min = 0,
    .sec = 0};


char datetime_str[256];
datetime_t Now_time;

// Arguments are received response message and the request's token
void handleData(ModbusMessage response, uint32_t token) 
{
    switch (response.getServerID())
    {
      case 1:
          Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
          for (auto& byte : response) {
            Serial.printf("%02X ", byte);
          }
          Serial.println("");
          break;
      case 2:
          Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", response.getServerID(), response.getFunctionCode(), token, response.size());
          for (auto& byte : response) {
            Serial.printf("%02X ", byte);
          }
          Serial.println("");
          break;

      default:
          break;
    }   
}

// Define an onError handler function to receive error responses
// Arguments are the error code returned and a user-supplied token to identify the causing request
void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  LOG_E("Client Error response: %02X - %s\n", (int)me, (const char *)me);
}

void setup()
{   
    // Init Serial monitor
    Serial.begin(115200UL);
    delay(5000); //Wait for initialization to succeed 

    Serial.println("__ OK __");
    
    Serial.println("__ Init RTC __");
    PCF85063A_Init();
    delay(500); //Wait for initialization to succeed
    Serial.println("__ RTC set time __"); 
    //PCF85063A_Set_All(Set_Time);

    Serial.println("__ Init Modbus __");

    //Set up Serial2 connected to Modbus RTU
    RTUutils::prepareHardwareSerial(MODBUS_SERIAL);
    MODBUS_SERIAL.begin(MODBUS_BAUD, MODBUS_CONFIG, MODBUS_RX, MODBUS_TX);

    // Set up ModbusRTU
    MBClient.onDataHandler(&handleData);
    MBClient.onErrorHandler(&handleError);
    MBClient.setTimeout(2000);
    MBClient.begin(MODBUS_SERIAL, 1);
    //Init modbus token
    MBToken = (uint32_t)0;
    data_request = false;
    data_ready = false;
    request = 0;

    next_request = millis();

    Serial.println("__ Setup done __");

}

uint32_t mbtoken(uint32_t token){
  if(token < 4294967295) token++;
  else token = 1;

  return token;
}

void loop()
{
    PCF85063A_Read_now(&Now_time);
    datetime_to_str(datetime_str,Now_time);
    //Serial.print("Now_time is " );
    //Serial.println(datetime_str);

    if (millis() - next_request > 10000) {
      // Issue the request
      data_request = true;
      
      MBToken = mbtoken(MBToken);
      err = MBClient.addRequest(MBToken, (uint8_t)1, READ_INPUT_REGISTER, (uint16_t)0, (uint16_t)4);
      if (err!=SUCCESS) {
        ModbusError e(err);
        LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
      }

      MBToken = mbtoken(MBToken);
      err = MBClient.addRequest(MBToken, (uint8_t)2, READ_INPUT_REGISTER, (uint16_t)0, (uint16_t)8);
      if (err!=SUCCESS) {
        ModbusError e(err);
        LOG_E("Error creating request: %02X - %s\n", (int)e, (const char *)e);
      }

      next_request = millis();
    }
}
```
and here is the print on the serial monitor

```
__ OK __
__ Init RTC __
__ RTC set time __
__ Init Modbus __
__ Setup done __
Response: serverID=1, FC=4, Token=00000001, length=11:
01 04 08 00 00 00 00 00 00 00 00 
[E] 17633| main.cpp             [  77] handleError: Client Error response: E0 - Timeout
Response: serverID=1, FC=4, Token=00000003, length=11:
01 04 08 00 00 00 00 00 00 00 00 
[E] 27633| main.cpp             [  77] handleError: Client Error response: E0 - Timeout
```

**Comment by Miq1** [comment]
Your code looks okay. If you will comment out the request for server 1, does server 2 respond in time?

**Reply by nicolouts** [reply]
Yes if I comment server 1. I get an answer from server 2

__ OK __
__ Init RTC __
__ RTC set time __
__ Init Modbus __
__ Setup done __
Response: serverID=2, FC=4, Token=00000001, length=19:
02 04 10 00 00 00 00 00 00 04 AF 00 00 00 00 00 00 00 00 
Response: serverID=2, FC=4, Token=00000002, length=19:
02 04 10 00 00 00 00 00 00 04 AF 00 00 00 00 00 00 00 00

**Comment by Miq1** [comment]
Ok, so we know the server is responding to requests at least. 
My next assumption would be communication speed. Try to gradually shorten the time between requests and still only request server 2. Is there a limit where you start to get timeouts?

**Comment by nicolouts** [comment]
I decreased request time until 5ms without any problem of timeout.
At 2ms, I receive erorr E8 - Request queue full

**Reply by Miq1** [reply]
Okay. The 2ms is due to the fact that the quiet time has to be maintained nd you will send requests faster than this time will allow.
But we learned that server 2 is fast enough by far.
Next we need to look into the details. Restore the requests to server 1, set the interval to 10000 again and compile your code with verbose logging. If you are using PlatformIO, you need to add ``-DLOG_LEVEL=6`` to the ``build_flags`` and compile again. The Arduino IDE has no such easy method, instead you will have to temporarily modify the file ``Logging.h`` to set the default logging level to LOG_LEVEL_VERBOSE right at the beginning of that file.
Please post the log output here for me to examine.

**Reply by nicolouts** [reply]
Here is the log

__ OK __
__ Init RTC __
__ RTC set time __
__ Init Modbus __
[D] 5620| ModbusClientRTU.cpp  [ 117] setTimeout: Timeout set to 2000
[V] 5620| RTUutils.cpp         [ 111] calculateInterval: Calc interval(115211)=1750
[D] 5620| ModbusClientRTU.cpp  [  91] doBegin: Client task 1070522748 started. Interval=1750
__ Setup done __
[V] 15621| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 15621| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/04
[D] 15621| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 15621| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[V] 15621| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 15621| ModbusClientRTU.cpp  [ 161] addRequestM: request for 02/04
[D] 15621| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 15622| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[D] 15622| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FCEE2FC/6:
  | 0000: 01 04 00 00 00 04                                 |......          |
[D] 15623| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[V] 15631| RTUutils.cpp         [ 261] receive: C/1755us without data after 13
[V] 15631| RTUutils.cpp         [ 271] receive: C/[V] Raw buffer received: @3FCEE388/13:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00 24 0D           |...........$.   |
[D] 15631| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 15631| ModbusClientRTU.cpp  [ 282] handleConnection: Data response (11 bytes) received.
[V] Data: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 15631| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
Response: serverID=1, FC=4, Token=00000001, length=11:
01 04 08 00 00 00 00 00 00 00 00 
[D] 15632| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FCEE2FC/6:
  | 0000: 02 04 00 00 00 08                                 |......          |
[D] 15633| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[D] 17634| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCEE2FC/1:
  | 0000: E0                                                |.               |
[D] 17634| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
[V] Data: @3FCEE2FC/1:
  | 0000: E0                                                |.               |
[D] 17634| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FCEE2FC/3:
  | 0000: 02 84 E0                                          |...             |
[E] 17635| main.cpp             [  77] handleError: Client Error response: E0 - Timeout
[V] 25623| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 25623| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/04
[D] 25623| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 25623| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[V] 25623| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 25623| ModbusClientRTU.cpp  [ 161] addRequestM: request for 02/04
[D] 25623| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 25623| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[D] 25624| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FCEE2FC/6:
  | 0000: 01 04 00 00 00 04                                 |......          |
[D] 25625| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[V] 25637| RTUutils.cpp         [ 261] receive: C/1753us without data after 13
[V] 25637| RTUutils.cpp         [ 271] receive: C/[V] Raw buffer received: @3FCEE388/13:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00 24 0D           |...........$.   |
[D] 25637| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 25637| ModbusClientRTU.cpp  [ 282] handleConnection: Data response (11 bytes) received.
[V] Data: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 25637| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FCEE328/11:
  | 0000: 01 04 08 00 00 00 00 00  00 00 00                 |...........     |
Response: serverID=1, FC=4, Token=00000003, length=11:
01 04 08 00 00 00 00 00 00 00 00 
[D] 25638| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FCEE2FC/6:
  | 0000: 02 04 00 00 00 08                                 |......          |
[D] 25639| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[D] 27640| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FCEE2FC/1:
  | 0000: E0                                                |.               |
[D] 27640| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
[V] Data: @3FCEE2FC/1:
  | 0000: E0                                                |.               |
[D] 27640| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FCEE2FC/3:
  | 0000: 02 84 E0                                          |...             |
[E] 27640| main.cpp             [  77] handleError: Client Error response: E0 - Timeout

**Comment by Miq1** [comment]
Unfortunately nothing special to see 😞 

If there would have been any bytes coming in they would have been logged. So it is a bit of trial and error now.

Can you switch the sequence of requests to add that for server 2 first, then that for server 1?

**Comment by nicolouts** [comment]
Same issue but now with server 1

__ OK __
__ Init RTC __
__ RTC set time __
__ Init Modbus __
__ Setup done __
Response: serverID=2, FC=4, Token=00000001, length=19:
02 04 10 00 00 00 00 00 00 04 B0 00 00 00 00 00 00 00 00 
[E] 17629| main.cpp             [  77] handleError: Client Error response: E0 - Timeout
Response: serverID=2, FC=4, Token=00000003, length=19:
02 04 10 00 00 00 00 00 00 04 B0 00 00 00 00 00 00 00 00

**Comment by Miq1** [comment]
Okay, next thing learned: it is neither server 1 nor server 2 responsible for the problem.

Next we will try if the servers can not recover fast enough after a message has passed the bus. In the ``begin`` call for the Modbus Client there is a third parameter to artificially increase the quiet time interval, the value is given in microseconds. Try with 10000 or even 20000 to see if the servers can manage both requests.

**Reply by nicolouts** [reply]
Sorry, I try 10000, 15000, 20000 and 30000 but unfortunally same error

**Comment by Miq1** [comment]
Too bad. There must be something special with your setup we do not yet see.

Several options:
- Do you have a spare ESP and RS485 module available? If so, we could configure it as a Modbus sniffer and log everything going on on the bus. 
- Second, we can set it up as another server and check if the effect also happens with this and one of your servers.
- Also helpful would be a scope connected to the RS485 wires to see if there are any spikes or such on the bus.

**Reply by nicolouts** [reply]
Hello Miq1,
Sorry for my latye answer.

I have done some test with another library and I'm able to read both server but I have some timeout also.
So it's difficult to find from where is the problem.
I have some doubt on my server 1, I don't have another server device. I need to buy another else to do the test.
Thank you, Regards
